sdks/go/pkg/beam/io/fhirio/search_test.go

```diff
@@ -0,0 +1,106 @@
+// limitations under the License.
+
+package fhirio
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/apache/beam/sdks/v2/go/pkg/beam"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/passert"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/ptest"
+)
+
+func TestSearch_Errors(t *testing.T) {
+	testCases := []struct {
+		name           string
+		client         fhirStoreClient
+		containedError string
+	}{
+		{
+			name:           "Search request returns error",
+			client:         requestReturnErrorFakeClient,
+			containedError: fakeRequestReturnErrorMessage,
+		},
+		{
+			name:           "Search request returns bad status",
+			client:         badStatusFakeClient,
+			containedError: strconv.Itoa(http.StatusForbidden),
+		},
+		{
+			name:           "Search request response body fails to be read",
+			client:         bodyReaderErrorFakeClient,
+			containedError: fakeBodyReaderErrorMessage,
+		},
+		{
+			name:           "Search request response body failed to be decoded",
+			client:         emptyResponseBodyFakeClient,
+			containedError: io.EOF.Error(),
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			p, s, testSearchQueryPCollection := ptest.CreateList([]SearchQuery{{}})
+			resources, failedSearches := search(s, "any", testSearchQueryPCollection, testCase.client)
+			passert.Empty(s, resources)
+			passert.Count(s, failedSearches, "", 1)
+			passert.True(s, failedSearches, func(errorMsg string) bool {
+				return strings.Contains(errorMsg, testCase.containedError)
+			})
+			pipelineResult := ptest.RunAndValidate(t, p)
+			err := validateCounter(pipelineResult, errorCounterName, 1)
+			if err != nil {
+				t.Fatalf("validateCounter returned error [%v]", err.Error())
+			}
+		})
+	}
+}
+
+func TestSearch_Pagination(t *testing.T) {
+	paginationFakeClient := &fakeFhirStoreClient{
+		fakeSearch: func(s, s2 string, m map[string]string, pageToken string) (*http.Response, error) {
+			if pageToken == "theNextPageToken" {
+				return &http.Response{
+					Body:       io.NopCloser(bytes.NewBufferString(`{"entry": [{"resource":{}}], "link": []}`)),
+					StatusCode: http.StatusOK,
+				}, nil
+			} else {
+				return &http.Response{
+					Body:       io.NopCloser(bytes.NewBufferString(`{"entry": [{"resource":{}}], "link": [{"relation":"next", "url":"https://healthcare.googleapis.com?_page_token=theNextPageToken"}]}`)),
+					StatusCode: http.StatusOK,
+				}, nil
+			}
+		},
+	}
+	p, s, testSearchQuery := ptest.CreateList([]SearchQuery{{}})
+	resourcesFound, failedSearches := search(s, "any", testSearchQuery, paginationFakeClient)
+	passert.Empty(s, failedSearches)
+	passert.Count(s, resourcesFound, "", 1)
+	resourcesFoundWithoutIdentifier := beam.DropKey(s, resourcesFound)
+	passert.True(s, resourcesFoundWithoutIdentifier, func(resourcesFound []string) bool {
+		return len(resourcesFound) == 2
```

Comment:
How is resources found equal to 2 when you don't return any resources in the fake client?

Reply:
Ah I see, the resource is empty.

Just to make this a bit more readable (in-case I miss the extra `{}`) could you make it something simple like

first page: `{"entry": [{"resource":{"resourceType": "Patient", "id": "1"}}], "link": []}`
second page: `{"entry": [{"resource":{"resourceType": "Patient", "id": "2"}}], "link": [{"relation":"next", "url":"https://healthcare.googleapis.com?_page_token=theNextPageToken"}]}`

Don't need to verify the contents.